File: day01/historian.py
import sys

def main():
    # parse the input
    list1 = []
    list2 = []

    with open(sys.argv[1]) as f:
        for line in f:
            toks = line.rstrip().split('   ')
            list1.append(int(toks[0]))
            list2.append(int(toks[1]))

    # Now solve it
    print('Part 1:', sum(map(lambda x: abs(x[0]-x[1]), zip(sorted(list1), sorted(list2)))))
    print('Part 2:', sum([x * len([y for y in list2 if y == x]) for x in list1]))
    
main()

File: day01/historian.pl
#!/usr/bin/env perl
use v5.40;
use List::Util qw(sum zip);

my (@list1, @list2);

while (<<>>) {
    chomp;
    my @toks = split '   ';
    push @list1, $toks[0];
    push @list2, $toks[1];
}

say "Part 1: ", sum map {abs $_->[0] - $_->[1]} zip [sort {$a<=>$b} @list1], [sort {$a<=>$b} @list2];

my $sum = 0;
for my $l1 (@list1) {
    $sum += $l1 * grep {$_ == $l1} @list2;
}
say "Part 2: $sum";
